[#15] Replace newlines with ZPL equivalent

itsmattking opened the pull request:
In FieldData, add an option to replace newlines `\n` with ZPL newline representation `\&`.

Also update FieldBlock extension to use this option so that any usage of fieldBlock will automatically replace newlines.

itsmattking committed:
[#15] Replace newlines with ZPL equivalent

In FieldData, add an option to replace newlines `\n`
with ZPL newline representation `\&`.

Also update FieldBlock extension to use this option
so that any usage of fieldBlock will automatically
replace newlines.

jbduncan commented on src/test/kotlin/com/sainsburys/k2zpl/command/FieldDataTest.kt:
```diff
@@ -23,5 +23,46 @@ class FieldDataTest : DescribeSpec({
+                """.trimIndent()
+                    fieldData(data = data, replaceNewlines = true)
+                }
+                result shouldBe "^FDsome data\\&with line breaks\\&\n"
```
Oh, can you clarify for me why the last `\n` is left alone?

In fact, I'm curious how this ever happens, because reading `FieldData`, it looks like either all newlines get replaced with `\&` or none do at all. 🤔

itsmattking replied:
This is because each command in ZPL is terminated by a newline, so it's required when adding a new command.

jbduncan replied:
Ah okay! I guess it must be happening somewhere else in the `k2zpl` DSL, so I'm happy with this. 👍

jbduncan commented on src/main/kotlin/com/sainsburys/k2zpl/command/FieldData.kt:
```diff
@@ -2,15 +2,33 @@ package com.sainsburys.k2zpl.command
-internal data class FieldData(val data: String) : ZplCommand {
+internal data class FieldData(val data: String, val replaceNewlines: Boolean = false) : ZplCommand {
     override val command: CharSequence = "^FD"
     override val parameters: LinkedHashMap<CharSequence, Any?> = linkedMapOf("d" to data)
```
It's just occurred to me that `parameters` should probably be of type `Map` on the `ZplCommand` interface so it becomes read-only and we can replace the implementation in the future if needed.

WDYT?

jbduncan replied:
If a mutable map is needed after all, I'd be tempted to still go for a `Map` but introduce a `addParameter` method so we don't expose all the mutation methods of the map and keep the interface small and easy to evolve.

itsmattking replied:
So the reason is I wanted to enforce the use of an ordered map in the subclasses of `ZplCommand` by using `LinkedHashMap`. Do you know if there's another way to achieve this?

jbduncan replied:
Not really, short of upgrading to Java 21 and using [sequenced collection interfaces](https://openjdk.org/jeps/431), if Android even supports them yet.

But the docs for `mapOf` and `mutableMapOf` say that they preserve entry order, and given they're used so often, I'd say this is a worthwhile tradeoff.

Also I'm a bit biased as I value immutability heavily over map iteration order and I'm used to maps having a non-deterministic iteration order in Golang. 😉 

WDYT?

jbduncan replied:
I suppose another way would be to introduce a new method `addParameters` that accepts a varargs of pairs and then adds them into an internal `LinkedHashMap`, with `override val parameters` becoming a non-override getter that is a read-only view into this internal `LinkedHashMap`, a la Java's `Collections::unmodifiableMap`.

jbduncan replied:
Yeah, actually the second option (`addParameters` and making `parameters` a read-only view) sounds like the way to go! It gives us the best of both worlds. ✨

itsmattking replied:
@jbduncan finally got around to this, let me know what you think 😄

jbduncan replied:
@itsmattking Sorry, I realise I wasn't clear! I've just made PR https://github.com/sainsburys-tech/k2zpl/pull/20 which shows what I have in mind.

jbduncan commented on src/main/kotlin/com/sainsburys/k2zpl/command/FieldData.kt:
```diff
@@ -2,15 +2,33 @@ package com.sainsburys.k2zpl.command
     override val parameters: LinkedHashMap<CharSequence, Any?> = linkedMapOf("d" to data)
+
+    override fun build(stringBuilder: StringBuilder): StringBuilder {
+        return with(stringBuilder) {
```
IIRC, there's no need for `with` because `StringBuilder.append` is a fluent chaining method, so we can replace it with:

```kotlin
stringBuilder
    .append(...)
    .append(...)
```

WDYT?

itsmattking replied:
Updated! 👍

jbduncan commented on src/main/kotlin/com/sainsburys/k2zpl/command/FieldBlock.kt:
```diff
@@ -82,6 +82,6 @@ fun ZplBuilder.fieldBlock(
         hangingIndent = hangingIndent
     )
-    fieldData(data)
+    fieldData(data, replaceNewlines = true)
```
I almost wonder if we can get away with making all `fieldData` usages replace newlines, and in turn eliminate this branch in the spirit of YAGNI.

WYDT?

itsmattking replied:
Good point, I don't see a time when we'd want to allow newlines specifically on this command type.

itsmattking replied:
Updated this one as well 👍

jbduncan reviewed (commented):
Just a few queries, but otherwise this PR looks pretty decent to me. 👍

itsmattking committed:
[#15] PR Feedback Updates

- Make newline replacement non-optional. It is very unlikely
we want to allow newline `\n` characters in field data as
this breaks ZPL output.
- Remove explicit use of `LinkedHashMap` as the `parameters` type
in favour of using just `Map`.
- Add helper `addParameters` to make explicit that we want a `LinkedHashMap`
to be used when adding parameters. This is to ensure key entry order is preserved.

itsmattking committed:
[#15] Use `addParameters` in tests as well, and update missed usages.
